server/services/v1/key_generator.go

```diff
@@ -0,0 +1,131 @@
+)
+
+const (
+	// generatorSubspaceKey is used to store ids in storage so that we can guarantee uniqueness
+	generatorSubspaceKey = "generator"
+	// int32IdKey is the prefix after generator subspace to store int32 counters
+	int32IdKey = "int32_id"
+	// timestampIdKey is the prefix after generator to store timestamps
+	timestampIdKey = "timestamp_id"
+)
+
+type generator struct {
+	txMgr *transaction.Manager
+}
+
+func newGenerator(txMgr *transaction.Manager) *generator {
+	return &generator{
+		txMgr: txMgr,
+	}
+}
+
+// get returns generated id for the supported primary key fields.
+func (g *generator) get(ctx context.Context, table []byte, field *schema.Field) ([]byte, error) {
+	switch field.Type() {
+	case schema.StringType, schema.UUIDType:
+		return []byte(fmt.Sprintf(`"%s"`, uuid.NewUUIDAsString())), nil
+	case schema.ByteType:
+		return jsoniter.Marshal([]byte(uuid.NewUUIDAsString()))
+	case schema.DateTimeType, schema.Int64Type, schema.Int32Type:
+		return g.generateInTx(ctx, table, field)
+	}
+	return nil, api.Errorf(codes.InvalidArgument, "unsupported type found in auto-generator")
+}
+
+// generateInTx is used to generate an id in a transaction. This is mainly used where there is no easy way to guarantee
+// uniqueness so what we are doing is reserving this id in storage before returning to the caller so that only one
+// id is assigned to one caller.
+func (g *generator) generateInTx(ctx context.Context, table []byte, field *schema.Field) ([]byte, error) {
+	for {
+		tx, err := g.txMgr.StartTx(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		var valueTS time.Time
+		var valueI64 int64
+		var valueI32 int32
+		switch field.Type() {
+		case schema.DateTimeType:
+			valueTS, _, err = g.generateTimestamp(ctx, tx, table)
+		case schema.Int64Type:
+			_, valueI64, err = g.generateTimestamp(ctx, tx, table)
+		case schema.Int32Type:
+			valueI32, err = g.generateInt(ctx, tx, table)
+		}
+		if err != nil {
+			_ = tx.Rollback(ctx)
+		}
+
+		if err = tx.Commit(ctx); err == nil {
+			if !valueTS.IsZero() {
+				return []byte(fmt.Sprintf(`"%s"`, valueTS.Format(time.RFC3339Nano))), nil
+			} else if valueI64 > 0 {
+				return []byte(fmt.Sprintf(`%d`, valueI64)), nil
+			} else if valueI32 > 0 {
+				return []byte(fmt.Sprintf(`%d`, valueI32)), nil
+			}
+		}
+		if err != kv.ErrConflictingTransaction {
+			return nil, err
+		}
+	}
+}
+
+// generateTimestamp stores values like this: "generator-<table>-timestamp_id-<tsNano>". The value is embedded in the
+// key to reduce the contention.
+// ToDo: cleanup old timestamps
+func (g *generator) generateTimestamp(ctx context.Context, tx transaction.Tx, table []byte) (time.Time, int64, error) {
+	tsTime := time.Now().UTC()
+	tsNano := time.Now().UTC().UnixNano()
+	key := keys.NewKey([]byte(generatorSubspaceKey), table, timestampIdKey, encoding.UInt32ToByte(uint32(tsNano)))
+
+	if err := tx.Insert(ctx, key, internal.NewTableData(encoding.UInt32ToByte(uint32(tsNano)))); err != nil {
```

Comment:
Why do we need to save the timestamp?

Reply:
That's ok. There will be duplicate error when we actually insert the document.
There will be any kind of clock skews between workers, so this check doesn't look protect anything.

Reply:
What about Replace? How will we be handling this in that case?

Reply:
Clock-skews are fine in this case - all we need is a guarantee to not re-generate the same id again.